reimplementation of Track classifiers and Track Collection mergers

cmsbuild opened the pull request:
I have fully reimplemented track selectors and merger.
Separated all concerns:
one class to use mva,
one class to use cuts (mimics mva output)
one class to merge selectors
one class to merge collections.
to be done
re-implement duplicate identification and merging
port to HLT

do not esclude that some non-main-stream use-cases are not covered yet.
some more cleanup and optimization is surely required.

I have also implemented a macro to instantiate a sort of DynArray as C++14 does not support anymore
dynamical array.

Full backward compatibility is not maintained.
1) 
The vertex identification uses only delta-z and not the track-vertex association:
the latter (besides being utterly show to verify with the current implementation) it is only working
for Iter0:  tracks from all other iterations are not associated to any vertex
2)
quality  looseSetWithPV and  highPuritySetWithPV is not set (nobody is using it)
and does not make much sense when using MVA anyhow (it can mark if the mva is prompt or detached, maybe.  TBD)
3) Iter5 and 6 are MVA only

In my opinion it requires still a couple of weeks of work.
I will not work for the next two weeks, so I prefer to submit this PR for reference and to collect feedback.

Automatically ported from CMSSW_7_5_X #9538 (original by @VinInn).

makortel committed:
Update validation campaigns

makortel committed:
Fix typos

makortel committed:
Add validation-style directory structure to trackingCompare.py

makortel committed:
Make canvas a local variable

makortel committed:
Change chi2 y axis to log scale

makortel committed:
Plot new histograms in vertex validation

VinInn committed:
new TrackMVAClassifier.cc

VinInn committed:
new TrackMVAClassifier.cc buildfile

VinInn committed:
add a simple merger of classifiers

VinInn committed:
name properly the cfi

VinInn committed:
fix hit iteration

VinInn committed:
add a CutClassifier as well

VinInn committed:
add a CutClassifier test

makortel committed:
Add support for ratio pad

makortel committed:
Draw the uncertainty of denominator, but redraw grid on top of that

makortel committed:
Allow ratio uncertainty to be disable per plot

makortel committed:
Denote ratio denominator uncertainty in legend

makortel committed:
Support rebinning

VinInn committed:
a simple dynamic array allocated on the stack

VinInn committed:
add multi-purpose TrackCollectionCloner

VinInn committed:
add a simple Collectin merger

VinInn committed:
start to code the dumplicate resolution

VinInn committed:
add test

VinInn committed:
add sequence to test

VinInn committed:
move test cff ot python dir

VinInn committed:
fix misstypes

VinInn committed:
fix order in destructor

VinInn committed:
simplify trajectory storing

VinInn committed:
simplify trajectory storing

VinInn committed:
clean

VinInn committed:
proper initialize associationmap

VinInn committed:
new tuple base cut-set

VinInn committed:
Merged TrackClassifiers from repository VinInn

VinInn committed:
merger should be backward compatible

VinInn committed:
Merged TrackClassifiers from repository VinInn

VinInn committed:
adapt to new mva

silviodonato committed:
../python/hltHiggsPostProcessors_cff.py

silviodonato committed:
../python/hltHiggsPostProcessors_cff.py

silviodonato committed:
HLTriggerOffline/Higgs/python/hltHiggsValidator_cfi.py

makortel committed:
Support new-style FastSim pileup DQM file naming

makortel committed:
Allow algoDirMap to be a function in addition to dict

makortel committed:
Monitor again FastSim pileup sample

makortel committed:
Comment out plotting of TrackingParticle for tracks, and non-PU TTbar and gen-vertex for vertices

They were not that useful to be plotted all the time.

makortel committed:
Draw ratio by default in validation scripts

makortel committed:
Add vertexCompare.py script for non-validation vertex plotting

VinInn committed:
move tokens in their own header file

VinInn committed:
new version of TrackClusterRemover

VinInn committed:
fix backward compatibility

VinInn committed:
Merged TrackClassifiers from repository VinInn

VinInn committed:
start porting configuration

VinInn committed:
ok till DuplicateListMerger

VinInn committed:
migration done to the point that reco runs

VinInn committed:
migration done to the point that reco runs

VinInn committed:
fix iter5 classifier

ferencek committed:
degridifying unpacked phi to break degeneracies in angular separations

VinInn committed:
all ok but qualitymask

VinInn committed:
fix eventcontent as well

alja committed:
Changes needed for ROOT6 on linux and osx release.

silviodonato committed:
Update hltHiggsValidator_cfi.py

silviodonato committed:
new path names after b-tag update

ferencek committed:
introduced a pseudo-random sign of the phi shift

venturia committed:
Protect against bunch crossing number = 3564

venturia committed:
Update cfg files to conddb v2

vkhristenko committed:
Temporary Fix to match Hcal Trig Towers

venturia committed:
New analysis to correlate the BPIX ladder positions with the beamspot position

venturia committed:
Added distributions of the bx and orbit number offsets

venturia committed:
Updated configurations

KiSooLee committed:
HLT name changed

venturia committed:
Added the description of some plugins

venturia committed:
Added the description of some plugins to analyze MC vertices

venturia committed:
Added the description of some configurations

safarzad committed:
add NoNoiseCleand MET trigger

safarzad committed:
Add NoNoiseClean Path

safarzad committed:
add New Path to module

safarzad committed:
add PFMET170_v

safarzad committed:
add PFMET170_v

safarzad committed:
add no noise path

safarzad committed:
add no noise path

venturia committed:
Added the description of the plugins to study the multiplicities

venturia committed:
fixed a carriage return

venturia committed:
fixed the instructions to load the library

cfmcginn committed:
Added endoffill to dqm for hi jet, photon, and fulltrack

mdhildreth committed:
add capability of saving Tracking Particles from pileup BX for PreMixing

mdhildreth committed:
Remove digi event content from PREMIX, PREMIXRAW, RAWSIM

cmsbuild commented:
A new Pull Request was created by @cmsbuild  for CMSSW_7_6_X.

reimplementation of Track classifiers and Track Collection mergers

It involves the following packages:

CommonTools/Utils
RecoLocalTracker/SubCollectionProducers
RecoMuon/Configuration
RecoTracker/Configuration
RecoTracker/ConversionSeedGenerators
RecoTracker/FinalTrackSelectors
RecoTracker/IterativeTracking
TrackingTools/PatternTools


@cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please review it and eventually sign? Thanks.
@ghellwig, @forthommel, @abbiendi, @nickmccoll, @threus, @venturia, @battibass, @makortel, @jhgoh, @jlagram, @cerati, @trocino, @rociovilar, @yduhm, @GiacomoSguazzoni, @rovere, @VinInn, @bellan, @mschrode, @dgulhan, @gbenelli, @appeltel, @gpetruc, @istaslis, @bachtis this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

cmsbuild committed:
Merge CMSSW_7_5_X into CMSSW_7_6_X.

rmanzoni committed:
fix d0-phi plot under AlcaBeamMonitor/Validation. Now both d0 and phi are computed wrt (0,0,0) and the correlation is preserved

slava77 commented:
@cmsbuild please test
(this failed compilation in the past 75X though; better register that issue here as well to continue review in 76X)

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: 8488a8da5e8c15d1f9ff5440f1424edec1ed7d12
 When I ran the RelVals I found an error in the following worklfows: 
 5.1 step1
<pre>runTheMatrix-results/5.1_TTbar+TTbarFS+HARVESTFS/step1_TTbar+TTbarFS+HARVESTFS.log

</pre>
135.4 step1
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>
140.53 step2
<pre>runTheMatrix-results/140.53_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI/step2_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI.log

</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9810/5641/summary.html

wddgit committed:
Replace getProcessParameterSet function

Replace the function getProcessParameterSet
in DQMServices. This function is being removed
from the CMSSW Framework because it does not work
properly with SubProcesses and causes issues
with multithreading. The replacement should
yield identical behavior to what was originally
there (except now it works in a SubProcess job).

cmsbuild committed:
Merge pull request #9766 from slava77/CMSSW_7_5_X_2015-06-19-1100/niPFrereco

 consistently use momentum at POCA for PFDisplacedVertex momentum computations (same as  #9703  from 74X)

wddgit committed:
Remove function getCentralityBinsFromDB

Remove function getCentralityBinsFromDB
because it uses getProcessParameterSet.
getProcessParameterSet is a Framework function
being removed from CMSSW because it does
not work in SubProcesses and causes issues
for multithreading. Note that
getCentralityBinsFromDB is being removed
instead of being repaired to use an alternative
for the following reasons.

First, nothing in the release calls it. The
only thing that references it is one commented
out line of code.

The function is in a DataFormats package and
accesses the EventSetup. This dependence is
not allowed.

The function is used in modules and pulls
configuration parameters from the top level
parameter set instead of from the parameter
set of the module that uses it. This is only
allowed in very special cases. If this function
was to be repaired in the future it should be
redesigned to get its configuration parameters
from the module in some way.

Also made related fixes to the header includes.

cmsbuild committed:
Merge pull request #9821 from rafaellopesdesa/FixOutOfBoundEnergyUncertaintyPhoton

Fix out of bound array in photon uncertainty calculation

makortel committed:
Add a script to produce validation plots from DQM files given as command line parameters

namapane committed:
Switch to new format for DT uncertainties

cmsbuild committed:
Merge pull request #9798 from makortel/recoTrackSelectorRefactor

Refactor RecoTrackSelector and RecoTrackRefSelector

cmsbuild committed:
Merge pull request #9794 from cms-btv-pog/CommonTMVAEvaluatorForBTagging_from-CMSSW_7_5_X_2015-06-18-1100

Switching b-taggers to a common TMVAEvaluator class

cmsbuild committed:
Merge CMSSW_7_5_X into CMSSW_7_6_X.

rmanzoni committed:
Added parameters to the migrad functor to supersede EDM defaults (new Minuit changed them after V5.32)

Commit:
remove outdated fastsim pileup option

Commit:
remove the genmix customisation function

Commit:
remove the pileupproducer package

Commit:
remove gen-mixing option from FamosProducer

Commit:
remove forgotten include file

Commit:
remove vertex smearing generators used in gen-mixing

davidlange6 committed:
Merge pull request #9744 from Dr15Jones/makeSimHitPrinterThreadSafe

Made file access in SimHitPrinter thread safe

davidlange6 committed:
Merge pull request #9745 from Dr15Jones/makeEcalSrFlagStaticThreadSafe

Made static EcalSrFlag::srfNames fully const

cmsbuild committed:
Merge pull request #9809 from ferencek/DegridifiedUnpackedPhi_from-CMSSW_7_5_X_2015-06-10-1100

Degridifying unpacked phi to break degeneracies in angular separations for packed candidates

wmtan committed:
Capabilities libraries are obsolete

alja committed:
Ignore LD_LIBRARY_PATH from the parent process.

alja committed:
Update m_source variable in checkFieldInfo(). The bug manifested in track-propagator field.

cschomak committed:
Include d<angle> vs. d<translation> plots in geometry comparison plotter

davidlange6 committed:
Merge pull request #9859 from lveldere/fastsim-removeGenMix-cmsDriver

FastSim: remove outdated pileup option from cmsDriver

cmsbuild committed:
Merge pull request #9865 from matteosan1/hltdqm_75X

Update to handle trigger paths with DiSC prefix

cmsbuild committed:
Merge pull request #9860 from lveldere/fastsim-removeGenMix-famosproducer

Fastsim: remove gen mixing from famosproducer

cmsbuild committed:
Merge pull request #9834 from MilanoBicocca-pix/CMSSW_7_6_X_fix_beamspot_d0_phi_dqm

Offline DQM: fix d0-phi plot under AlcaBeamMonitor/Validation

cmsbuild committed:
Merge pull request #9817 from silviodonato/DQMWHbb

DQM for HLT_Ele27_WPLoose_Gsf_WHbbBoost_v1 - 75X

cmsbuild committed:
Merge pull request #9805 from silviodonato/DQM-higgs-btag-fix

Fix of path names in HLT Higgs validation code - 75X

cmsbuild committed:
Merge pull request #9804 from silviodonato/DQMbtagfix

Fix path name in HLT BTV valdiation code

makortel committed:
Allow dEdx plots be ignored for FastSim (don't exist anymore in 750pre4 for FS)

cmsbuild committed:
Merge pull request #9753 from cfmcginn/HI50nsDQMUpdate

Added endoffill to dqm for hi jet, photon, and fulltrack

cmsbuild committed:
Merge pull request #9749 from barvic/cscdqm_run2_histos_patch_75X

CSC DQM Run2 histograms adjustments and fixes for CMSSW_7_5_X

cmsbuild committed:
Merge pull request #9872 from wmtan/NoCapabilities

Capabilities libraries are obsolete

cmsbuild committed:
Merge pull request #9774 from safarzad/myPath

add no noise clean path to SUSYBSM_inclusiveMET_cff.py  & SUSYBSM_postProcessor_cff.py & SusyExoValidation_cff.py

danduggan committed:
online DQM updates in 76x: EvD configs and small fixes

cmsbuild committed:
Merge pull request #9883 from namapane/DTSwitchToNewRecoUncert_76X

Switch to new format for DT uncertainties

cmsbuild committed:
Merge pull request #9795 from vkhristenko/tpfix_75

Temporary Fix to match Hcal Trig Towers

Commit:
GEMCSCSegAlgoRR: initialize xe and ye

    DataFormats/GeometryVector/interface/extBasic3DVector.h:78:14: error:
    'ye' may be used uninitialized in this function
    [-Werror=maybe-uninitialized]
         v{x,y,z,w}{}
              ^
    RecoLocalMuon/GEMCSCSegment/plugins/GEMCSCSegAlgoRR.cc:217:14: note:
    'ye' was declared here
        float xe, ye, ze = 0.0;

Signed-off-by: David Abdurachmanov <David.Abdurachmanov@cern.ch>

cmsbuild committed:
Merge CMSSW_7_5_X into CMSSW_7_6_X.

alja committed:
Update filed info in FF.

alja committed:
Reset field estimate on new data file.

davidlange6 committed:
Merge pull request #9896 from davidlt/fix-xe-and-ye-GEMCSCSegAlgoRR

GEMCSCSegAlgoRR: initialize xe and ye

cmsbuild committed:
Merge pull request #9891 from danduggan/EvDinput_76x

online DQM updates in 76x: EvD configs and small fixes

cmsbuild committed:
Merge pull request #9882 from makortel/trackingValidation_v2

Update tracking MC validation scripts

cmsbuild committed:
Merge pull request #9875 from alja/75-field

75x Fireworks: port bugfix in mag field determination from 74

cmsbuild committed:
Merge pull request #9870 from wddgit/removegetCentralityBinsFromDB

Remove function getCentralityBinsFromDB

cmsbuild committed:
Merge pull request #9869 from wddgit/getProcessParameterSetDQMServices

Replace getProcessParameterSet function

davidlange6 committed:
Merge pull request #9772 from Dr15Jones/uninitValueJetDeltaRValueMapProducer-7_5

Fix an uninitialized value in JetDeltaRValueMapProducer

cmsbuild committed:
Merge pull request #9858 from MilanoBicocca-pix/CMSSW_7_6_X_Minuit_EDM_Fix

BeamSpot Fitter: added parameters to the migrad functor to supersede EDM defaults

cmsbuild committed:
Merge pull request #9776 from ngrenz/CMSSW_7_5_X_ngrenz_string_makeMaps

reduce creation of string objects and use char* instead of string

cmsbuild committed:
Merge pull request #9741 from cms-l1t-offline/l1t-FED-vector-HF-uppgrade-bugfix-75x

L1t fed vector hf uppgrade bugfix 75x

cmsbuild committed:
Merge CMSSW_7_5_X into CMSSW_7_6_X.

mmusich committed:
two liner to make AlignmentTrackSelectorModule multi-threaded friendly

mmusich committed:
additional fixes

davidlange6 committed:
Merge pull request #9888 from davidlange6/mem_reduct_744p1

fixing memory leak in L1 CSCTrackFinder

davidlange6 committed:
Merge pull request #9773 from venturia/dpganalysis_sistriptools-75x

Fix for orbit number = 3564, update of the documentation and other improvements

cmsbuild committed:
Merge pull request #9793 from KiSooLee/HLT_DQM_75X

HLT name changed

cmsbuild committed:
Merge CMSSW_7_5_X into CMSSW_7_6_X.

deguio committed:
minor fixes online applications. same as #9623

alja committed:
Fix module name for Jet collection.

alja committed:
Change Jet collection module.

cmsbuild committed:
Merge pull request #9927 from dmitrijus/dqm_not_init_75x

Dqm not init 75x

cmsbuild committed:
Merge pull request #9910 from mmusich/fixAlignmentTrackSelectorMultiThreadedFriendly76X

Fix alignment track selector multi threaded friendly for 76X (fwdport of #9906)

cmsbuild committed:
Merge CMSSW_7_5_X into CMSSW_7_6_X.

davidlange6 committed:
Merge pull request #9737 from mdhildreth/TPUpdatesForPreMixing

TrackingParticle updates for pre mixing

cmsbuild committed:
Merge pull request #9930 from alja/75x-config

75x Fireworks: port Fireworks configuration changes from 74

cmsbuild committed:
Merge pull request #9759 from mmarionncern/METPAT_METCor75X

New MET correction/uncertainty tool

cmsbuild committed:
Merge CMSSW_7_5_X into CMSSW_7_6_X.

davidlange6 committed:
remove forgotten usage of fastsim pileup

davidlange6 committed:
Merge pull request #9942 from davidlange6/removeFSPU

remove forgotten usage of fastsim pileup

cmsbuild committed:
Merge CMSSW_7_5_X into CMSSW_7_6_X.

cmsbuild committed:
Merge CMSSW_7_5_X into CMSSW_7_6_X.

cmsbuild committed:
Merge pull request #9923 from deguio/fixOnline75x

minor fixes online applications. same as #9623

cmsbuild committed:
Merge pull request #9879 from cschomak/CMSSW_7_6_X_GeometryComparison

Include d<angle> vs. d<translation> plots in geometry comparison plotter

cmsbuild committed:
Merge CMSSW_7_5_X into CMSSW_7_6_X.

slava77 commented:
@VinInn  ping
[I can't recall if you are back already; if not, it's OK to wait]

VinInn commented:
I plan to fix FastSim and HI
My understanding is that FastSim is broken in IB, so I have problem in testing....

davidlange6 commented:
fastsim is ok again in yesterdays IB..

> On Jun 30, 2015, at 8:31 AM, Vincenzo Innocente <notifications@github.com> wrote:
> 
> I plan to fix FastSim and HI
> My understanding is that FastSim is broken in IB, so I have problem in testing....
> 
> —
> Reply to this email directly or view it on GitHub.
>

VinInn committed:
Merged TrackClassifiers from repository VinInn

VinInn commented:
a Problem with fastSim PU.
Its is a reco file and I have redefined the event-content for MVA
so it needs to be regenerated....

VinInn committed:
fix fastsim

cmsbuild commented:
Pull request #9810 was updated. @smuzaffar, @civanch, @diguida, @lveldere, @cvuosalo, @ssekmen, @mdhildreth, @monttj, @cmsbuild, @alja, @franzoni, @Dr15Jones, @cerminar, @deguio, @slava77, @danduggan, @vadler, @mmusich, @davidlange6 can you please check and sign again.

lveldere commented:
would you mind rebasing?

If there are any FastSim questions,
feel free to ask.

Lukas

slava77 commented:
-1 need a rebase

VinInn committed:
fix HI as well

cmsbuild commented:
Pull request #9810 was updated. @smuzaffar, @civanch, @diguida, @lveldere, @cvuosalo, @ssekmen, @mdhildreth, @monttj, @cmsbuild, @alja, @franzoni, @Dr15Jones, @cerminar, @deguio, @slava77, @danduggan, @vadler, @mmusich, @davidlange6 can you please check and sign again.

VinInn commented:
instead of rebase made new PR #9990

VinInn commented:
please close

slava77 commented:
-1
superseded

alja commented:
+1

lveldere commented:
-1

VinInn commented:
@lveldere 
sorry, what is your suggestion to progress? I can remove the MVA from the output for good actually: nobody is using it at the moment at it is also of difficult use as its value depends on the iteration

lveldere commented:
Actually MVA could be removed from the mixing indeed.
I implemented a while ago to try something out,
but that never worked out, and I doubt it will ever be needed.

(btw, I think floats is the friendly class name you are looking for, just
in case)

On Thu, Jul 2, 2015 at 8:39 AM, Vincenzo Innocente <notifications@github.com
> wrote:

> @lveldere <https://github.com/lveldere>
> sorry, what is your suggestion to progress? I can remove the MVA from the
> output for good actually: nobody is using it at the moment at it is also of
> difficult use as its value depends on the iteration
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9810#issuecomment-117924722>.
>
